File: frontend/lib/services/export.ts
/**
 * Export Service - Frontend API client for CSV/ZIP exports
 */

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'

export type ExportType = 'keywords' | 'competitors' | 'visibility' | 'mentions' | 'prompts' | 'sentiment' | 'all'

export interface ExportOptions {
    days?: number  // For visibility and mentions exports
    includeHistory?: boolean  // For keywords export
}

export const exportService = {
    /**
     * Export keywords data as CSV
     */
    async exportKeywords(brandId: string, includeHistory: boolean = false): Promise<Blob> {
        const url = `${API_URL}/api/export/${brandId}/keywords?include_history=${includeHistory}`
        return await downloadBlob(url)
    },

    /**
     * Export competitors data as CSV
     */
    async exportCompetitors(brandId: string): Promise<Blob> {
        const url = `${API_URL}/api/export/${brandId}/competitors`
        return await downloadBlob(url)
    },

    /**
     * Export visibility history as CSV
     */
    async exportVisibility(brandId: string, days: number = 90): Promise<Blob> {
        const url = `${API_URL}/api/export/${brandId}/visibility?days=${days}`
        return await downloadBlob(url)
    },

    /**
     * Export brand mentions as CSV
     */
    async exportMentions(brandId: string, days: number = 90): Promise<Blob> {
        const url = `${API_URL}/api/export/${brandId}/mentions?days=${days}`
        return await downloadBlob(url)
    },

    /**
     * Export tracked prompts as CSV
     */
    async exportPrompts(brandId: string): Promise<Blob> {
        const url = `${API_URL}/api/export/${brandId}/prompts`
        return await downloadBlob(url)
    },

    /**
     * Export sentiment analysis history as CSV
     */
    async exportSentiment(brandId: string): Promise<Blob> {
        const url = `${API_URL}/api/export/${brandId}/sentiment`
        return await downloadBlob(url)
    },

    /**
     * Export all data as ZIP file
     */
    async exportAll(brandId: string): Promise<Blob> {
        const url = `${API_URL}/api/export/${brandId}/all`
        return await downloadBlob(url)
    },

    /**
     * Generic export function
     */
    async export(brandId: string, type: ExportType, options?: ExportOptions): Promise<Blob> {
        let url = `${API_URL}/api/export/${brandId}/${type}`

        const params = new URLSearchParams()
        if (options?.days) params.append('days', options.days.toString())
        if (options?.includeHistory) params.append('include_history', 'true')

        if (params.toString()) {
            url += `?${params.toString()}`
        }

        return await downloadBlob(url)
    },

    /**
     * Download export and trigger browser download
     */
    async downloadExport(brandId: string, type: ExportType, options?: ExportOptions): Promise<void> {
        const blob = await this.export(brandId, type, options)

        const filename = generateFilename(type)
        triggerDownload(blob, filename)
    }
}

/**
 * Helper to download a blob from URL
 */
async function downloadBlob(url: string): Promise<Blob> {
    const response = await fetch(url, {
        credentials: 'include'
    })

    if (!response.ok) {
        throw new Error(`Export failed: ${response.statusText}`)
    }

    return await response.blob()
}

/**
 * Generate filename based on export type
 */
function generateFilename(type: ExportType): string {
    const date = new Date().toISOString().split('T')[0]

    if (type === 'all') {
        return `mentha_export_${date}.zip`
    }

    return `mentha_${type}_${date}.csv`
}

/**
 * Trigger browser download of blob
 */
function triggerDownload(blob: Blob, filename: string): void {
    const url = window.URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = filename
    document.body.appendChild(a)
    a.click()
    window.URL.revokeObjectURL(url)
    document.body.removeChild(a)
}

File: backend/app/api/endpoints/export.py
"""
Export API Endpoints

Endpoints for exporting Mentha data in CSV and ZIP formats.
"""

from fastapi import APIRouter, HTTPException, Depends
from fastapi.responses import Response
from typing import Optional
from datetime import datetime

from app.api.deps import get_current_user_id
from app.services.export_service import get_export_service

router = APIRouter(prefix="/export", tags=["Export"])


@router.get("/{brand_id}/keywords")
async def export_keywords(
    brand_id: str,
    include_history: bool = False,
    user_id: str = Depends(get_current_user_id)
):
    """Export keywords data as CSV."""
    service = get_export_service()
    csv_content = await service.export_keywords(brand_id, include_history)
    
    if not csv_content:
        raise HTTPException(status_code=404, detail="No keywords found or export failed")
    
    filename = f"keywords_{datetime.utcnow().strftime('%Y%m%d')}.csv"
    headers = service.generate_csv_response_headers(filename)
    
    return Response(
        content=csv_content,
        media_type="text/csv",
        headers=headers
    )


@router.get("/{brand_id}/competitors")
async def export_competitors(
    brand_id: str,
    user_id: str = Depends(get_current_user_id)
):
    """Export competitors data as CSV."""
    service = get_export_service()
    csv_content = await service.export_competitors(brand_id)
    
    if not csv_content:
        raise HTTPException(status_code=404, detail="No competitors found or export failed")
    
    filename = f"competitors_{datetime.utcnow().strftime('%Y%m%d')}.csv"
    headers = service.generate_csv_response_headers(filename)
    
    return Response(
        content=csv_content,
        media_type="text/csv",
        headers=headers
    )


@router.get("/{brand_id}/visibility")
async def export_visibility(
    brand_id: str,
    days: int = 90,
    user_id: str = Depends(get_current_user_id)
):
    """Export AI visibility history as CSV."""
    service = get_export_service()
    csv_content = await service.export_visibility_history(brand_id, days)
    
    if not csv_content:
        raise HTTPException(status_code=404, detail="No visibility data found or export failed")
    
    filename = f"visibility_{datetime.utcnow().strftime('%Y%m%d')}.csv"
    headers = service.generate_csv_response_headers(filename)
    
    return Response(
        content=csv_content,
        media_type="text/csv",
        headers=headers
    )


@router.get("/{brand_id}/mentions")
async def export_mentions(
    brand_id: str,
    days: int = 90,
    user_id: str = Depends(get_current_user_id)
):
    """Export brand mentions as CSV."""
    service = get_export_service()
    csv_content = await service.export_mentions(brand_id, days)
    
    if not csv_content:
        raise HTTPException(status_code=404, detail="No mentions found or export failed")
    
    filename = f"mentions_{datetime.utcnow().strftime('%Y%m%d')}.csv"
    headers = service.generate_csv_response_headers(filename)
    
    return Response(
        content=csv_content,
        media_type="text/csv",
        headers=headers
    )


@router.get("/{brand_id}/prompts")
async def export_prompts(
    brand_id: str,
    user_id: str = Depends(get_current_user_id)
):
    """Export tracked prompts as CSV."""
    service = get_export_service()
    csv_content = await service.export_prompt_tracking(brand_id)
    
    if not csv_content:
        raise HTTPException(status_code=404, detail="No prompts found or export failed")
    
    filename = f"prompts_{datetime.utcnow().strftime('%Y%m%d')}.csv"
    headers = service.generate_csv_response_headers(filename)
    
    return Response(
        content=csv_content,
        media_type="text/csv",
        headers=headers
    )


@router.get("/{brand_id}/sentiment")
async def export_sentiment(
    brand_id: str,
    user_id: str = Depends(get_current_user_id)
):
    """Export sentiment analysis history as CSV."""
    service = get_export_service()
    csv_content = await service.export_sentiment_analysis(brand_id)
    
    if not csv_content:
        raise HTTPException(status_code=404, detail="No sentiment data found or export failed")
    
    filename = f"sentiment_{datetime.utcnow().strftime('%Y%m%d')}.csv"
    headers = service.generate_csv_response_headers(filename)
    
    return Response(
        content=csv_content,
        media_type="text/csv",
        headers=headers
    )


@router.get("/{brand_id}/all")
async def export_all(
    brand_id: str,
    user_id: str = Depends(get_current_user_id)
):
    """Export all data as a ZIP file containing multiple CSVs."""
    service = get_export_service()
    zip_content = await service.export_all_as_zip(brand_id)
    
    if not zip_content:
        raise HTTPException(status_code=500, detail="Export failed")
    
    filename = f"mentha_export_{datetime.utcnow().strftime('%Y%m%d')}.zip"
    headers = service.generate_zip_response_headers(filename)
    
    return Response(
        content=zip_content,
        media_type="application/zip",
        headers=headers
    )
